s3fs/core.py

```diff
@@ -908,27 +965,17 @@ async def _info(self, path, bucket=None, key=None, refresh=False, version_id=Non
             return {"name": path, "size": 0, "type": "directory"}
         if key:
             try:
-                out = await self._call_s3(
-                    "head_object",
-                    self.kwargs,
-                    Bucket=bucket,
-                    Key=key,
-                    **version_id_kw(version_id),
-                    **self.req_kw,
-                )
-                return {
-                    "ETag": out["ETag"],
-                    "Key": "/".join([bucket, key]),
-                    "LastModified": out["LastModified"],
-                    "Size": out["ContentLength"],
-                    "size": out["ContentLength"],
-                    "name": "/".join([bucket, key]),
-                    "type": "file",
-                    "StorageClass": "STANDARD",
-                    "VersionId": out.get("VersionId"),
-                }
-            except FileNotFoundError:
-                pass
+                if self.version_aware:
+                    out = await self._version_aware_info(path, version_id)
+                else:
+                    try:
+                        out = await self._simple_info(path)
+                    except PermissionError:
+                        # If the permissions aren't enough for scanning a prefix
+                        # then fall back to using normal HEAD_OBJECT
+                        out = await self._version_aware_info(path, version_id)
+                if out:
+                    return out
```

Comment:
So, is `out` ever `None` or empty here, or would that have raised by now? This is the end of the function anyway, so the `if` seems redundant. It would be nice to specify what this method can be expected to return - indeed, I should have done this.

Reply:
It can be None indeed, if it fall backs to the third (implicit) branch.

Reply:
OK, got it. Thanks for clarifying - for some reason I found it hard to follow that. That all sounds good, maybe you want to add a comment in the code - or not.